src/utils/middleware/csrf.middleware.js

```diff
@@ -0,0 +1,79 @@
+const csrf = require('csrf')
+
+/**
+ * @param logger {Logger}
+ * @param sessionKey {string}
+ * @param secretKey {string}
+ * @param tokenKey {string}
+ */
+const configureCsrfMiddleware = (logger, sessionKey, secretKey = 'csrfSecret', tokenKey = 'csrfToken') => {
+  logger.debug('--- CSRF middleware configuration ---')
+  logger.debug(`Secret is set at req.[${sessionKey}][${secretKey}]`)
+  logger.debug(`Token is checked at req.body|query[${tokenKey}]`)
+  logger.debug('-------------------------------------')
+  /**
+   * @param csrfToken {string}
+   * @param csrfSecret {string}
+   * @param req {e.Request}
+   */
+  const csrfValid = (csrfToken, csrfSecret, req) => {
+    if (!csrfSecret) {
+      logger.debug('CSRF secret not found when validating token')
+      return false
+    }
+    if (!['put', 'post'].includes(req.method.toLowerCase())) {
+      return true
+    }
+    return csrf().verify(csrfSecret, csrfToken)
+  }
+
+  /**
+   * @param req {e.Request}
+   * @param res {e.Response}
+   * @param next {e.NextFunction}
+   */
+  const setSecret = (req, res, next) => {
+    const csrfSecret = req[sessionKey][secretKey]
+    if (!csrfSecret) {
+      logger.debug('Synchronising CSRF secret')
+      req[sessionKey][secretKey] = csrf().secretSync()
+    }
+    next()
+  }
+
+  /**
+   * @param req {e.Request}
+   * @param res {e.Response}
+   * @param next {e.NextFunction}
+   */
+  const checkToken = (req, res, next) => {
+    const csrfSecret = req[sessionKey][secretKey]
+    const csrfToken = (req.body && req.body[tokenKey]) || (req.query && req.query[tokenKey])
```

Comment:
Should it be possible to pass the csrf token in a query paramter? since the validation is only enforced on PUT or POST operations I think it should only ever be expected in the body, unless there's a case where we are passing in a query parameter